src/libdreddtest/src/mutation_replace_unary_operator_test.cc

```diff
@@ -0,0 +1,209 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "libdredd/mutation_replace_unary_operator.h"
+
+#include <memory>
+#include <string>
+
+#include "clang/AST/Expr.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/ASTMatchers/ASTMatchers.h"
+#include "clang/ASTMatchers/ASTMatchersInternal.h"
+#include "clang/Basic/Diagnostic.h"
+#include "clang/Basic/SourceManager.h"
+#include "clang/Frontend/ASTUnit.h"
+#include "clang/Rewrite/Core/RewriteBuffer.h"
+#include "clang/Rewrite/Core/Rewriter.h"
+#include "clang/Tooling/Tooling.h"
+#include "libdreddtest/gtest.h"
+#include "llvm/ADT/SmallVector.h"
+
+namespace dredd {
+namespace {
+
+void TestReplacement(const std::string& original, const std::string& expected,
+                     int num_replacements,
+                     const std::string& expected_dredd_declaration) {
+  auto ast_unit = clang::tooling::buildASTFromCodeWithArgs(original, {"-w"});
+  ASSERT_FALSE(ast_unit->getDiagnostics().hasErrorOccurred());
+  auto function_decl = clang::ast_matchers::match(
+      clang::ast_matchers::functionDecl(clang::ast_matchers::hasName("foo"))
+          .bind("fn"),
+      ast_unit->getASTContext());
+  ASSERT_EQ(1, function_decl.size());
+
+  auto unary_operator = clang::ast_matchers::match(
+      clang::ast_matchers::unaryOperator().bind("op"),
+      ast_unit->getASTContext());
+  ASSERT_EQ(1, unary_operator.size());
+
+  MutationReplaceUnaryOperator mutation(
+      *unary_operator[0].getNodeAs<clang::UnaryOperator>("op"));
+
+  clang::Rewriter rewriter(ast_unit->getSourceManager(),
+                           ast_unit->getLangOpts());
+  int mutation_id = 0;
+  std::unordered_set<std::string> dredd_declarations;
+  mutation.Apply(ast_unit->getASTContext(), ast_unit->getPreprocessor(), 0,
+                 mutation_id, rewriter, dredd_declarations);
+  ASSERT_EQ(num_replacements, mutation_id);
+  ASSERT_EQ(1, dredd_declarations.size());
+  ASSERT_EQ(expected_dredd_declaration, *dredd_declarations.begin());
+
+  const clang::RewriteBuffer* rewrite_buffer = rewriter.getRewriteBufferFor(
+      ast_unit->getSourceManager().getMainFileID());
+  std::string rewritten_text(rewrite_buffer->begin(), rewrite_buffer->end());
+  ASSERT_EQ(expected, rewritten_text);
+}
+
+TEST(MutationReplaceUnaryOperatorTest, MutateMinus) {
+  std::string original = "void foo() { -2; }";
+  std::string expected =
+      "void foo() { __dredd_replace_unary_operator_Minus_int([&]() -> int& { "
+      "return "
+      "static_cast<int&>(2); }, "
```

Comment:
This seems wrong: does it make sense to cast the literal value 2 to an int reference?

Reply:
Changed to only use static casts when it is not a constant.